### PR TITLE
maven plugin generates api-tests and model-tests in generated-test-sources

### LINKED ISSUE #10917: [REQ] [JAVA] maven plugin generates test in generated-test-sources
### Is your feature request related to a problem? Please describe.

Maven uses two different folders generated-sources and generated-test-sources. Test related files supose to get into generated-test-source folder. In current implementation all generated files in a single forlder - ${project.build.directory}/generated-sources/openapi by default. 

there are problems with the approach: 
 - test files go to production artifact
 - production artifact becomes dependencies (it increses dangerous in the light of recent vulnerability in logging library - the more non essential libraries in prod the likely to catch new vulnerability)

I suspect that majority of java developers just disable generating the test files

I wonder if it it possible to make mvn plugin alligned with conventional folder-structure 

## Describe the solution you'd like

The solution is to introduce additional parameter testOutput which is ${project.build.directory}/generated-test-sources/openapi by default.

## Describe alternatives you've considered

there are several workarounds:
 - copy test files after generation with another maven plugin 
 - use org.codehouse.mojo:build-helper-maven-plugin to mark the folder with test files as mvn-test-source

both of these ways have different flaws

### PR DESCRIPTION
**Pull request by feech**
the change makes the maven plugin to generate Model-tests and API-tests in generated-test-sources
closes #10917

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.
@nmuesch @bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) 
@wing328 @Dever4eg @zhuangheihei

### CODE REVIEW

**Comment by feech** [comment]
target the issue
https://github.com/OpenAPITools/openapi-generator/issues/10917

**Comment by feech** [comment]
Hello @nmuesch @bbdouglas @sreeshas  @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog @lwlee2608 @wing328 @Dever4eg @zhuangheihei.
Could anyone please review this?
I'm happy to update the PR if some things don't match expectations.

**Comment by feech** [comment]
Hello
@nmuesch @bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10)
@wing328 @Dever4eg @zhuangheihei @jimschubert

might I draw your attention to the PR

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenConfig.java [hunk 1]
```diff
@@ -66,6 +66,8 @@
 
     String outputFolder();
 
+    String testOutputFolder();
```
Comment (spacether) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenConfig.java:69]: Why have this and the getter?

Comment (feech) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenConfig.java:69]: runtime files tree and test files tree could be in different folders.
ex. in Java there are different folders:
${project.build.directory}/generated-sources and ${project.build.directory}/generated-test-sources

currently test files get into ${project.build.directory}/generated-sources


there are 2 main issues with having test files in prod artifact:
1. test files go to production artifact
2. production code gets test dependencies

I suspect that currently majority of java developers just disable generating the test files

Comment (feech) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenConfig.java:69]: I am sorry,  @spacether. It seems I didn't get your question. Both - outputFolder and testOutputFolder are getters. But `outputFolder` returns folder for a prod code, but `testOutputFolder` for tests

Comment (spacether) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenConfig.java:69]: What is the difference between getTestOuputFolder and testOutputDir?

Comment (spacether) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenConfig.java:69]: If one of them is a path can we put path in its name

**Comment by feech** [comment]
@spacether 
I also created another PR which aims to resolve initial issue - #10917
the other PR -  https://github.com/OpenAPITools/openapi-generator/pull/11294
in other solution - i made it with adding additional-parameter instead of main-parameter.
so the other solution has smaller scope and will not impact other generators.

**A note for all reviewers**: any of this two solutions solve the issue for java-generators and I intend to merge only one of them - or maybe there is 3rd solution which I will do based on your feedback

**Comment by feech** [comment]
@spacether
Is there anything else which prevents the PR from merging into master?

**Comment by wing328** [comment]
Closed via https://github.com/OpenAPITools/openapi-generator/pull/11294

**Pull request closed without merging**